### PR TITLE
CMake: Silent `git diff` after the build, one-click GUI-based config, much cleanup & fixes

### PR DESCRIPTION
**Pull request by tschw**
See commits for details. Hope you like.

Maybe should also bump the CMake version to 3.3, because that's the version introducing `--format=zip` as used for packaging `pyluxcoretools`... I left it alone for now.

### CODE REVIEW

**Comment by tschw** [comment]
Yep, let me fix that.

> Doesn't it happen on Windows too (you have to run an OpenCL rendering to get the error because the code is compiled at runtime) ?

Most probably does. 

It had gotten super late, once had the PR together, so I didn't do runtime tests, especially since there's CI in place (but looking closer, I notice that it does CPU-only testing). I will also check on whether those CMake warnings come from something I've touched.

I guess the actual reason for CI failure
```
The job exceeded the maximum time limit for jobs, and has been terminated.
```
is due to load (daytime in the USA).

**Comment by Dade916** [comment]
```
I guess the actual reason for CI failure

The job exceeded the maximum time limit for jobs, and has been terminated.

is due to load (daytime in the USA).
```

Yeah, TravisCI is nearly useless at the moment, I have reduce the number of unit tests executed there but the combo of slow VM and too much intensive unit tests hits the time limit very often.

**Comment by tschw** [comment]
> The very last string is missing a "\n"

In-place edit of commit `Re-implemented kernel preprocessor with CMake`:
```patch
diff --git a/cmake/Scripts/PreprocessKernel.cmake b/cmake/Scripts/PreprocessKernel.cmake
index 98518b4d..dc9c1ff1 100644
--- a/cmake/Scripts/PreprocessKernel.cmake
+++ b/cmake/Scripts/PreprocessKernel.cmake
@@ -9,6 +9,6 @@ string(REGEX REPLACE "\"" "\\\\\"" CL_SOURCE_CODE "${CL_SOURCE_CODE}")

 # Wrap lines in string literals:
 string(REGEX REPLACE "\r?\n$" "" CL_SOURCE_CODE "${CL_SOURCE_CODE}")
-string(REGEX REPLACE "\r?\n" "\\\\n\"\n\"" CL_SOURCE_CODE "\"${CL_SOURCE_CODE}\"")
+string(REGEX REPLACE "\r?\n" "\\\\n\"\n\"" CL_SOURCE_CODE "\"${CL_SOURCE_CODE}\\n\"")

 configure_file(${template} ${DST} @ONLY)
```

As for the warnings, one is from `FindOpenMP.cmake` shipped with CMake versions below 3.10.
It only happens when targeting CMake version < 3.1.

https://gitlab.kitware.com/cmake/cmake/issues/17292

The other is from my code, because `pyside-uic` is not found. Is it installed on the build host? If so, where to best look for it on Linux?

> Yeah, TravisCI is nearly useless at the moment, I have reduce the number of unit tests executed there but the combo of slow VM and too much intensive unit tests hits the time limit very often.

Maybe could gain a bit of headroom for testing by saving build time by using `ninja`...  It builds roughly twice as fast as MSBuild, which is already quite a bit faster than GNU-Make, I think.

File: cmake/Dependencies.cmake [hunk 1]
```diff
@@ -43,9 +43,14 @@ if(NOT APPLE)
     find_package(PNG REQUIRED)
     include_directories(BEFORE SYSTEM ${PNG_PNG_INCLUDE_DIR})
 	# Find Python Libraries
-	find_package(PythonLibs)
+	find_package(PythonLibs 3.5)
 endif()
 
+find_program(PYSIDE_UIC NAME pyside-uic
+		HINTS "${PYTHON_INCLUDE_DIRS}/../Scripts"
```
Comment (tschw) [inline cmake/Dependencies.cmake]: Will that hint work for all platforms?

Comment (Dade916) [reply cmake/Dependencies.cmake]: Yup, pyside-uic is in system path on linux by default.

**Comment by Dade916** [comment]
Thanks Toby, it works fine here, I have merged the pull request.

**Comment by tschw** [comment]
This was really just scratching the tip of the iceberg. There is much more that could be done on the CMake end, such as declaring installation rules to generate the SDK in a cross-platform fashion, or an automated, external build for the dependencies (that would allow me to easily pull up to VisualStudio 2017 and get rid of yet another version of that massive software package requiring gigabytes of disk space).

Using [target exports](https://cmake.org/Wiki/CMake/Tutorials/Exporting_and_Importing_Targets) it becomes unnecessary to directly fiddle with find scripts, paths and libraries: Instead you just have your projects depend on targets generated by external builds as if they were local ones.

I've become bit of a "CMake power user" over the recent years (it's how I stay halfway sane targeting both UNIX and Windows) and help with that once it fits into our schedule.

**Comment by Dade916** [comment]
Yes but keep in mind that, given the limited resources, we have a bit to prioritize our work. Any change to the building process requires to test about 10+ different builds: CPU-only/With OpenCL support, Linux/Windows, Stand-alone/BlendLuxCore/SDK/Python wheel builds, static linking/DLLs ... it is a lot of work so we usually esitate to change building process once we have something working.

Something I would like to have (and use) is separate directories for DEBUG and RELEASE builds. At the moment, when I switch from RELEASE to DEBUG, I have to recompile everything while, with different directories, I would have a way to reuse, at least some object from a previous compilation.

However a change like this would require separate DEBUG/RELEASE bin and lib directories and it would impact all the scripts for packing the aforementioned releases. Even a small change in this area lead to a lot of (annoying) testing.

**Pull request merged**